tron/trondaemon.py

```diff
@@ -195,27 +211,66 @@ def _run_mcp(self):
 
     def _run_reactor(self):
         """Run the twisted reactor."""
-        threading.Thread(
-            target=reactor.run,
-            daemon=True,
-            kwargs=dict(installSignalHandlers=0)
-        ).start()
         signal_map = {
             signal.SIGHUP: self._handle_reconfigure,
-            signal.SIGINT: self._handle_shutdown,
+            signal.SIGINT: self._sigint_handler,
             signal.SIGTERM: self._handle_shutdown,
             signal.SIGQUIT: self._handle_shutdown,
             signal.SIGUSR1: self._handle_debug,
         }
+        signal.pthread_sigmask(signal.SIG_BLOCK, signal_map.keys())
+
+        threading.Thread(
+            target=reactor.run,
+            daemon=True,
+            kwargs=dict(installSignalHandlers=0)
+        ).start()
+
         while True:
-            signal.pthread_sigmask(signal.SIG_BLOCK, signal_map.keys())
-            signum = signal.sigwait(set(signal_map.keys()))
-            logging.info("Got signal %s" % signum)
+            try:
+                # We use a sigtimedwait instead of a sigwait here because in the
+                # event other threads try to interrupt the main thread, a
+                # KeyboardInterrupt will be thrown. A sigwait will not unblock,
+                # but a sigtimedwait will.
+                signum = signal.sigtimedwait(set(signal_map.keys()), 0)
+                if signum is not None:
+                    signum = signal.Signals(signum.si_signo)
+            except KeyboardInterrupt:
+                signum = signal.SIGINT
+
             if signum in signal_map:
+                logging.info(f"Got signal {str(signum)}")
                 signal_map[signum](signum, None)
 
+    def _make_sigint_handler(self, prev_handler=None):
+        """ Creates a SIGINT handler that takes into account a previous
+        handler to differentiate between a user request to shutdown, versus
+        another source we want to prevent from interrupting the reactor.
+
+        :type prev_handler: function
+        :param prev_handler: The previous SIGINT handler, set by another source.
+                             We use it to verify whether or not a SIGINT we
+                             received is a genuine shutdown request.
+        """
+        def handler(signum, frame):
+            try:
+                if prev_handler is not None:
+                    prev_handler(signum, frame)
+            except KeyboardInterrupt:
+                # Previous signal handler didn't raise another exception,
+                # so must be user requesting shutdown.
+                pass
+            except Exception as e:
+                # We received a SIGINT, but was caused by another thread
+                # aborting due to its own error. In this case, we don't want to
+                # stop running.
+                log.info(f"Non-reactor thread raised: {e}")
```

Comment:
Can you make this an error?

Reply:
Oops! I thought I had already done this. I'll make sure it happens this time, @qui.